gufe/storage/resultsclient.py

```diff
@@ -0,0 +1,141 @@
+from typing import Any
+
+from gufe.storage.resultstore import ResultStore
+from gufe.storage.metadatastore import JSONMetadataStore
+
+
+class _ResultContainer:
+    """
+    Abstract class, represents all data under some level of the heirarchy.
+    """
+    def __init__(self, parent, path_component):
+        self.parent = parent
+        self._path_component = self._to_path_component(path_component)
+        self._cache = {}
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, self.__class__)
+            and self.path == other.path
+        )
+
+    @staticmethod
+    def _to_path_component(item: Any) -> str:
+        """Convert input (object or string) to path string"""
+        if isinstance(item, str):
+            return item
+
+        # TODO: instead of str(hash(...)), this should return the digest
+        # that is being introduced in another PR; Python hash is not stable
+        # across sessions
+        return str(hash(item))
+
+    def __getitem__(self, item):
+        hash_item = self._to_path_component(item)
+
+        if hash_item not in self._cache:
+            self._cache[hash_item] = self._load_next_level(item)
+
+        return self._cache[hash_item]
```

Comment:
I remember from discussion we talked about the possibility of this checking if the file existed and only proceeding to `load_next_level` if `item` was a dir

Reply:
Added this. Gotta say, I feel like I have a pretty good design when adding functionality like this is as easy as it was.